lib/api/core/validation/index.js

```diff
@@ -0,0 +1,788 @@
+    }
+
+    if (!fieldSpec.multivalued.value && (fieldSpec.multivalued.hasOwnProperty('minCount'))) {
+      throw new InternalError('"minCount" is not valid when multivalued field is disabled.');
+    }
+
+    if (!fieldSpec.multivalued.value && (fieldSpec.multivalued.hasOwnProperty('maxCount'))) {
+      throw new InternalError('"maxCount" is not valid when multivalued field is disabled.');
+    }
+
+    if (fieldSpec.multivalued.hasOwnProperty('minCount') && fieldSpec.multivalued.hasOwnProperty('maxCount') && fieldSpec.multivalued.minCount > fieldSpec.multivalued.maxCount) {
+      throw new InternalError('"minCount" can not be greater than "maxCount".');
+    }
+  }
+
+  return true;
+};
+
+/**
+ * @param {string} indexName
+ * @param {string} collectionName
+ * @param {*} validatorFilter
+ * @param {boolean} dryRun
+ * @returns {Promise}
+ */
+Validation.prototype.curateValidatorFilter = function (indexName, collectionName, validatorFilter, dryRun) {
+  var
+    promise,
+    sandboxDsl = new Dsl(),
+    query = {
+      bool: {
+        must: validatorFilter
+      }
+    };
+
+  promise = sandboxDsl.register(indexName, collectionName, query);
+
+  if (!dryRun) {
+    promise.then(() => {
+      return this.dsl.register(indexName, collectionName, query);
+    });
+  }
+
+  return promise;
+};
+
+/**
+ * @param {ValidationType} validationType
+ * @throws {PluginImplementationError}
+ */
+Validation.prototype.addType = function (validationType) {
+  if (!validationType.typeName) {
+    throw new PluginImplementationError('The typeName property must be defined in the validation type object.');
+  }
+
+  if (!validationType.validate || typeof validationType.validate !== 'function') {
+    throw new PluginImplementationError(`The type ${validationType.typeName} must implement the function 'validate'.`);
+  }
+
+  if (!validationType.validateFieldSpecification || typeof validationType.validateFieldSpecification !== 'function') {
+    throw new PluginImplementationError(`The type ${validationType.typeName} must implement the function 'validateFieldSpecification'.`);
+  }
+
+  if (validationType.hasOwnProperty('allowChildren') && validationType.allowChildren) {
+    if (!validationType.getStrictness || typeof validationType.getStrictness !== 'function') {
+      throw new PluginImplementationError(`The allowing children type ${validationType.typeName} must implement the function 'getStrictness'.`);
+    }
+
+    this.typeAllowsChildren.push(validationType.typeName);
+  }
+
+  if (this.types[validationType.typeName]) {
+    throw new PluginImplementationError(`The type ${validationType.typeName} is already defined.`);
+  }
+
+  this.types[validationType.typeName] = validationType;
+};
+
+/**
+ * @param {*} object
+ * @param {string[]} allowedProperties
+ * @returns {boolean}
+ */
+function checkAllowedProperties(object, allowedProperties) {
+  var result = true;
+
+  if (typeof object !== 'object') {
```

Comment:
Should also check if object is not null and not an array, as `typeof` returns `object` on both these cases